modules/nsq/src/it/scala/com/snowplowanalytics/snowplow/enrich/nsq/test/Containers.scala

```diff
@@ -0,0 +1,274 @@
+/*
+ * Copyright (c) 2023 Snowplow Analytics Ltd. All rights reserved.
+ *
+ * This program is licensed to you under the Apache License Version 2.0,
+ * and you may not use this file except in compliance with the Apache License Version 2.0.
+ * You may obtain a copy of the Apache License Version 2.0 at http://www.apache.org/licenses/LICENSE-2.0.
+ *
+ * Unless required by applicable law or agreed to in writing,
+ * software distributed under the Apache License Version 2.0 is distributed on an
+ * "AS IS" BASIS, WITHOUT WARRANTIES OR CONDITIONS OF ANY KIND, either express or implied.
+ * See the Apache License Version 2.0 for the specific language governing permissions and limitations there under.
+ */
+
+package com.snowplowanalytics.snowplow.enrich.nsq.test
+
+import org.slf4j.LoggerFactory
+
+import cats.implicits._
+
+import cats.effect.{Resource, Sync, Async, ContextShift, Blocker}
+
+import org.http4s.client.{JavaNetClientBuilder, Client => Http4sClient}
+import org.http4s.{Request ,Method, Uri}
+
+import org.testcontainers.containers.{BindMode, GenericContainer => JGenericContainer, Network}
+import org.testcontainers.containers.wait.strategy.Wait
+import org.testcontainers.containers.output.Slf4jLogConsumer
+
+import com.dimafeng.testcontainers.{FixedHostPortGenericContainer, GenericContainer}
+
+import com.snowplowanalytics.snowplow.enrich.nsq.generated.BuildInfo
+
+object Containers {
+
+  case class NetworkInfo(
+    networkAlias: String,
+    broadcastAddress: String,
+    httpPort: Int,
+    tcpPort: Int
+  )
+
+  case class NetworkTopology(
+    lookup1: NetworkInfo,
+    nsqd1: NetworkInfo,
+    lookup2: NetworkInfo,
+    nsqd2: NetworkInfo,
+    sourceTopic: String,
+    goodDestTopic: String,
+    badDestTopic: String
+  )
+
+  /**
+   * Integration test tries to access nsqd from outside of docker network
+   * therefore it can't access it with docker service name.
+   * Since nsqd broadcasts its address through nsqlookup, it is impossible to
+   * broadcast its address as localhost and docker service name at the same time
+   * therefore we've created second nsqd and nsqlookup instances that will be used by
+   * integration tests and NSQ messages sent to first nsqd instance will be replicated
+   * to second nsqd instance with nsq_to_nsq tool.
+   */
```

Comment:
The problem is lookup operation in NSQ. NSQ consumer connects to [nsqlookupd](https://github.com/snowplow/enrich/blob/enrich-nsq/modules/nsq/src/main/scala/com/snowplowanalytics/snowplow/enrich/nsq/Source.scala#L89) to discover nsqd instances. And [nsqd](https://nsq.io/components/nsqd.html) broadcasts its address through nsqlookup. Then NSQ consumer tries to connect to the nsqd client through brodcasted address. The problem in here is that NSQ applications and Enrich-NSQ is running on same Docker network. However, tests are running outside of that network. Therefore, if we make the nsqd broadcast address `localhost`, Enrich NSQ consumer in the Docker container can't access to it because they aren't sharing same localhost. It can only access to nsqd via Docker network alias. In order to solve this problem, we can make the nsqd broadcast address Docker network alias. In that case however, integration tests can't access to nsqd because they are not in Docker network.

In order to solve this problem, I've created two nsqd instances. The broadcast address of the one of them is `localhost`. And other's broadcast address is its Docker network alias. Messages are relayed one to another with nsq_to_nsq. With this way, Enrich-NSQ in Docker container can use the nsqd with Docker network alias. Then, its content will be relayed to nsqd instance with localhost as broadcast address. Tests will consume messages from this nsqd instance.

I know this is quite complicated and solution isn't straighforward. But this was the only solution I can find. Please let me know if you have something better in mind. Also, we can have 1-1 if you have still some questions.

Reply:
Thanks for the detailed explanations, I get it now! I can't think of an easier alternative.